### PR TITLE
DFReader.py: correct python3 support

### PR DESCRIPTION
**Pull request by khancyr**
Correct DFReader problem on python 3

### CODE REVIEW

**Comment by khancyr** [comment]
I will do more testing later, but it works for .BIN and tlog . I didn't find a way to generate text DF log ... 
We should definitly add a test data suite

File: DFReader.py [hunk 1]
```diff
@@ -505,9 +518,9 @@ class DFReader_binary(DFReader):
     def __init__(self, filename, zero_time_base=False):
         DFReader.__init__(self)
         # read the whole file into memory for simplicity
-        f = open(filename, mode='rb')
-        self.data = f.read()
-        self.data_len = len(self.data)
+        with open(filename, 'rb') as f:
+            self.data = f.read()
+            self.data_len = len(self.data)
         f.close()
```
Comment (peterbarker) [inline DFReader.py]: Kill this line

File: DFReader.py [hunk 2]
```diff
@@ -588,15 +601,14 @@ def _parse_next(self):
...
         
         return m
 
 def DFReader_is_text_log(filename):
     '''return True if a file appears to be a valid text log'''
-    f = open(filename)
-    ret = (f.read(8000).find('FMT, ') != -1)
+    with open(filename, 'r') as f:
+        ret = (f.read(8000).find('FMT, ') != -1)
     f.close()
```
Comment (peterbarker) [inline DFReader.py]: Kill this line

File: DFReader.py [hunk 3]
```diff
@@ -605,8 +617,8 @@ class DFReader_text(DFReader):
     def __init__(self, filename, zero_time_base=False):
         DFReader.__init__(self)
         # read the whole file into memory for simplicity
-        f = open(filename, mode='r')
-        self.lines = f.readlines()
+        with open(filename, 'r') as f:
+            self.lines = f.readlines()
         f.close()
```
Comment (peterbarker) [inline DFReader.py]: Kill this line

Comment (peterbarker) [inline DFReader.py]: Running the same log through with python2 vs python3 is giving different results:
```
-2016-11-02 01:11:40.72: LOCAL_POSITION_NED {time_boot_ms : 350433, x : 1.2293812036514282, y : 0.10330259054899216, z : -30.021739959716797, vx : -0.19804619252681732, vy : 0.17870385944843292, vz : -0.5509760975837708}
+2016-11-02 01:11:40.72: LOCAL_POSITION_NED {time_boot_ms : 350433, x : 1.22938120365, y : 0.103302590549, z : -30.0217399597, vx : -0.198046192527, vy : 0.178703859448, vz : -0.550976097584}
```

Comment (peterbarker) [inline DFReader.py]:
```
pbarker@bluebottle:~/rc/logs$ python3 /home/pbarker/.local/bin/mavlogdump.py df/jmillwood/7c36f55e6569d7942ff9b6f961a6d8fba6a8d60e.bin >/tmp/b
Traceback (most recent call last):
  File "/home/pbarker/.local/bin/mavlogdump.py", line 4, in <module>
    __import__('pkg_resources').run_script('pymavlink==2.0.8', 'mavlogdump.py')
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 719, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1504, in run_script
    exec(code, namespace, namespace)
  File "/home/pbarker/.local/lib/python3.5/site-packages/pymavlink-2.0.8-py3.5-linux-x86_64.egg/EGG-INFO/scripts/mavlogdump.py", line 55, in <module>
    zero_time_base=args.zero_time_base)
  File "/home/pbarker/.local/lib/python3.5/site-packages/pymavlink-2.0.8-py3.5-linux-x86_64.egg/pymavlink/mavutil.py", line 1220, in mavlink_connection
    from pymavlink import DFReader
  File "/home/pbarker/.local/lib/python3.5/site-packages/pymavlink-2.0.8-py3.5-linux-x86_64.egg/pymavlink/DFReader.py", line 35, in <module>
    "q": ("q", None, long),  # Backward compat
NameError: name 'long' is not defined
pbarker@bluebottle:~/rc/logs$ 
```

Comment (peterbarker) [inline DFReader.py]: It also looks like mavlogdump.py is separately broken for .log files, so I can't test py2 vs py3 for that.

I'll look at this problem.

Comment (peterbarker) [inline DFReader.py]: @khancyr Actually, sorry, mavlogdump.py is working just fine on master for dumping text-based logs.

This is the exception under python2:

```
pbarker@bluebottle:~/rc/pymavlink((HEAD detached at khancyr/python_correction1))$ mavlogdump.py ~/crash2.log
Traceback (most recent call last):
  File "/home/pbarker/.local/bin/mavlogdump.py", line 4, in <module>
    __import__('pkg_resources').run_script('pymavlink==2.0.8', 'mavlogdump.py')
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 719, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 1504, in run_script
    exec(code, namespace, namespace)
  File "/home/pbarker/.local/lib/python2.7/site-packages/pymavlink-2.0.8-py2.7-linux-x86_64.egg/EGG-INFO/scripts/mavlogdump.py", line 58, in <module>
    zero_time_base=args.zero_time_base)
  File "/home/pbarker/.local/lib/python2.7/site-packages/pymavlink-2.0.8-py2.7-linux-x86_64.egg/pymavlink/mavutil.py", line 1232, in mavlink_connection
    m = DFReader.DFReader_text(device, zero_time_base=zero_time_base)
  File "/home/pbarker/.local/lib/python2.7/site-packages/pymavlink-2.0.8-py2.7-linux-x86_64.egg/pymavlink/DFReader.py", line 628, in __init__
    self.init_clock()
  File "/home/pbarker/.local/lib/python2.7/site-packages/pymavlink-2.0.8-py2.7-linux-x86_64.egg/pymavlink/DFReader.py", line 377, in init_clock
    m = self.recv_msg()
  File "/home/pbarker/.local/lib/python2.7/site-packages/pymavlink-2.0.8-py2.7-linux-x86_64.egg/pymavlink/DFReader.py", line 456, in recv_msg
    return self._parse_next()
  File "/home/pbarker/.local/lib/python2.7/site-packages/pymavlink-2.0.8-py2.7-linux-x86_64.egg/pymavlink/DFReader.py", line 689, in _parse_next
    self._add_msg(m)
  File "/home/pbarker/.local/lib/python2.7/site-packages/pymavlink-2.0.8-py2.7-linux-x86_64.egg/pymavlink/DFReader.py", line 483, in _add_msg
    self.flightmode = mavutil.mode_string_acm(m.Mode)
  File "/home/pbarker/.local/lib/python2.7/site-packages/pymavlink-2.0.8-py2.7-linux-x86_64.egg/pymavlink/mavutil.py", line 1683, in mode_string_acm
    return "Mode(%u)" % mode_number
TypeError: %u format: a number is required, not unicode
pbarker@bluebottle:~/rc/pymavlink((HEAD detached at khancyr/python_correction1))$ which python
```

And under python3:
```
pbarker@bluebottle:~/rc/pymavlink((HEAD detached at khancyr/python_correction1))$ python3 /home/pbarker/.local/bin/mavlogdump.py ~/crash2.log
Traceback (most recent call last):
  File "/home/pbarker/.local/bin/mavlogdump.py", line 4, in <module>
    __import__('pkg_resources').run_script('pymavlink==2.0.8', 'mavlogdump.py')
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 719, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1504, in run_script
    exec(code, namespace, namespace)
  File "/home/pbarker/.local/lib/python3.5/site-packages/pymavlink-2.0.8-py3.5-linux-x86_64.egg/EGG-INFO/scripts/mavlogdump.py", line 55, in <module>
    zero_time_base=args.zero_time_base)
  File "/home/pbarker/.local/lib/python3.5/site-packages/pymavlink-2.0.8-py3.5-linux-x86_64.egg/pymavlink/mavutil.py", line 1227, in mavlink_connection
    from pymavlink import DFReader
  File "/home/pbarker/.local/lib/python3.5/site-packages/pymavlink-2.0.8-py3.5-linux-x86_64.egg/pymavlink/DFReader.py", line 35, in <module>
    "q": ("q", None, long),  # Backward compat
NameError: name 'long' is not defined
pbarker@bluebottle:~/rc/pymavlink((HEAD detached at khancyr/python_correction1))$ 
```

Comment (peterbarker) [inline DFReader.py]: Note that this is not just the timestamp which is different; the precision is also different.

tridge can say if it's OK or not.

Comment (khancyr) [inline DFReader.py]: @peterbarker yep ! sorry didn't have time to finish this before, I will look at precision this week !

**Comment by khancyr** [comment]
@peterbarker , Ok I check the precision problem. The cause it the way float are print in python2 not the precision itself, and on that point python3 is more precise.
Do I really need to change the way it is print in python3 ? Because the only solution I found is ugly as only use format like this : ```` python
             if(isinstance(msg.param_value, float)):
                msg.param_value = "{:.12f}".format(msg.param_value)
````
will make all value 4.0 like 4.000000000000 ....

File: DFReader.py [hunk 4]
```diff
@@ -8,10 +8,12 @@
 Partly based on SDLog2Parser by Anton Babushkin
 '''
 from __future__ import print_function
+from past.builtins import long
```
Comment (cclauss) [inline DFReader.py]: A [different solution](https://github.com/ArduPilot/pymavlink/commit/463b63c8339d09a7c7479fe6f383c047450b24df) to line 11 was just checked into master.  The rest of this PR is very important for achieving Python 3 functionality.

File: DFReader.py [hunk 5]
```diff
@@ -36,6 +38,12 @@
     "Q": ("Q", None, long),  # Backward compat
     }
 
+def u_ord(c):
+    if sys.hexversion >= 0x03000000:
```
Comment (cclauss) [inline DFReader.py]: Perhaps `return ord(c) if sys.version_info.major < 3 else c`

**Comment by amilcarlucas** [comment]
How actual is this PR ? It looks like a couple of people are using python3 on master now.

**Comment by cclauss** [comment]
Status on this one?  Seems a bit out of date.

**Comment by amilcarlucas** [comment]
@khancyr can you rebase this ?

**Comment by peterbarker** [comment]
Well, probably unrelated to this PR, but:

```
pbarker@bluebottle:~/rc/pymavlink(python_correction1)$ mavlogdump.py $TLOG 
Traceback (most recent call last):
  File "/home/pbarker/.local/bin/mavlogdump.py", line 4, in <module>
    __import__('pkg_resources').run_script('pymavlink==2.2.8', 'mavlogdump.py')
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 742, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1503, in run_script
    exec(code, namespace, namespace)
  File "/home/pbarker/.local/lib/python3.6/site-packages/pymavlink-2.2.8-py3.6-linux-x86_64.egg/EGG-INFO/scripts/mavlogdump.py", line 58, in <module>
    zero_time_base=args.zero_time_base)
  File "/home/pbarker/.local/lib/python3.6/site-packages/pymavlink-2.2.8-py3.6-linux-x86_64.egg/pymavlink/mavutil.py", line 1257, in mavlink_connection
    return mavserial(device, baud=baud, source_system=source_system, source_component=source_component, autoreconnect=autoreconnect, use_native=use_native)
  File "/home/pbarker/.local/lib/python3.6/site-packages/pymavlink-2.2.8-py3.6-linux-x86_64.egg/pymavlink/mavutil.py", line 803, in __init__
    self.port = serial.Serial(self.device, 1200, timeout=0,
AttributeError: module 'serial' has no attribute 'Serial'
pbarker@bluebottle:~/rc/pymavlink(python_correction1)$ 
```

**Comment by peterbarker** [comment]
Actually, dumping a tlog with Python3 on master does seem to work, sooo...

**Comment by cclauss** [comment]
http://pyserial.readthedocs.io/en/latest/shortintro.html is _not_ part of the Python standard library so __python3 -m pip install pyserial__ might be required.

**Comment by peterbarker** [comment]
@amilcarlucas The question is "why doesn't it die on master" :-)

I've merged this.

**Pull request closed without merging**